Client/trunk/ParaEngineClient/BlockEngine/RenderableChunk.cpp

```diff
@@ -805,24 +887,58 @@ namespace ParaEngine
 				//--------------------------------------------------------------
 				BlockVertexCompressed* pBlockModelVertices = NULL;
 				unprocessedInstCount--;
-				int32 nFaceCount = tessellator.TessellateBlock(pChunk, instanceGroup[inst], dwShaderID, &pBlockModelVertices);
-				if (nFaceCount > 0)
+				if (materialId < 0)
 				{
-					int32 nVertexCount = nFaceCount * 4;
-					if (nFreeFaceCountInVertexBuffer >= nFaceCount)
+					int32 nFaceCount = tessellator.TessellateBlock(pChunk, instanceGroup[inst], dwShaderID, &pBlockModelVertices);
+					if (nFaceCount > 0)
 					{
-						memcpy(pVertices, pBlockModelVertices, sizeof(BlockVertexCompressed)*nVertexCount);
-						pVertices += nVertexCount;
-						vertexOffset += nVertexCount;
+						int32 nVertexCount = nFaceCount * 4;
+						if (nFreeFaceCountInVertexBuffer >= nFaceCount)
+						{
+							memcpy(pVertices, pBlockModelVertices, sizeof(BlockVertexCompressed)*nVertexCount);
+							pVertices += nVertexCount;
+							vertexOffset += nVertexCount;
 
-						nFaceCountCompleted += nFaceCount;
-						pTask->AddRectFace(nFaceCount);
-						nFreeFaceCountInVertexBuffer -= nFaceCount;
+							nFaceCountCompleted += nFaceCount;
+							pTask->AddRectFace(nFaceCount);
+							nFreeFaceCountInVertexBuffer -= nFaceCount;
+						}
+						else
+						{
+							// this could happen when block changes when we are still processing it. 
+							OUTPUT_LOG("fatal error: not enough face count in vertex buffer. \n");
+						}
 					}
-					else
+				}
+				else
+				{
+					pBlockModelVertices = pTemplate->GetBlockModelByData(nBlockData).GetVertices();
```

Comment:
refactor this code to final state. 

we need to refactor this code into a special tesselator function, and remove hidden faces and calculate AO as well for cube and maybe other (stair , slope) shapes in future.